File: backtrack/intermediate/generate_parentheses.py
class Solution:
    # @param an integer
    # @return a list of string
    def generateParenthesis(self, n):
        ret = []
        self.helper(n, n, ret, "")
        return ret

    def helper(self, l, r, ret, str):
        if l == 0 and r == 0:
            ret.append(str)
        else:
            if l > 0:
                str += '('
                self.helper(l-1, r, ret, str)
                str = str[:-1]
            if l < r:
                str += ')'
                self.helper(l, r-1, ret, str)
                str = str[:-1]

sol = Solution()
print sol.generateParenthesis(3)

File: various/ruby/intermediate/generate_parentheses.rb
# https://leetcode.com/problems/generate-parentheses/
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  @result = []
  generate(n, n, [])
  @result
end

def generate(l, r, pats)
  if l == 0 && r == 0
    @result << pats.join
  else
    if l > 0
      generate(l - 1, r, pats << '(')
      pats.pop
    end
    if l < r
      generate(l, r - 1, pats << ')')
      pats.pop
    end
  end
end

puts generate_parenthesis(4)
